### PR TITLE
Command Context Inter-Dependencies

### PR DESCRIPTION
**Pull request by Narvin-Chana**
- Adds a new way to have dependencies between different gpu operations.
- The concept of a "Frame" no longer exists in Vex, instead we allow for frame-like behavior, scoped by calls to Present().
- CommandContexts now have two different submission modes: DeferToPresent and Immediate. The first will attempt to defer the submission (if it is advantageous to do so) until just before the next present. Immediate instantly submits the command context to the GPU.
- Added a tests file with a synchronization torture test to allow for finding edge cases in our sync code.
- Submitting immediately now gives a SyncToken that can be used to wait for the work to be done manually.
- Deferred submission instead uses the present chain to know when to wait on the CPU.
- Reworked the Vulkan submission logic due to the present no longer being mandatory, the logic is that we now NEVER expose the backbuffer to the user, instead we render to (FRAME_BUFFERING) PresentTextures, whose contents are copied to the current backbuffer when the user calls present. This effectively decouples general graphics work, and acquisition/presenting!
- Removed the fact that RHITextureState is a FLAGS enum. DX12 actually does not support multiple simultaneous states, identical to Vulkan so this was adding complexity for no gain.
- Other smaller fixes/changes to remove some synchronisation edge cases or crashes that could occur.

### CODE REVIEW

File: src/Vex/GfxBackend.cpp [hunk 1]
```diff
@@ -115,73 +118,125 @@ GfxBackend::~GfxBackend()
...
+    rhi.WaitForTokenOnCPU(presentTokens[currentFrameIndex]);
 
-    rhi.SubmitAndPresent(queuedCommandLists, *swapChain, currentFrameIndex, isFullscreenMode);
-    queuedCommandLists.clear();
+    // Before presenting we have to handle all the queued for submission command lists (and their dependencies).
+    SubmitDeferredWork();
+
+    // Open a new command list that will be used to copy the presentTexture to the backbuffer, and presenting.
+    RHITexture& presentTexture = GetRHITexture(GetCurrentPresentTexture().handle);
+    RHITexture backBuffer = swapChain->AcquireBackBuffer(currentFrameIndex);
```
[the remark this answers is not in the document]
Comment (Narvin-Chana) [reply src/Vex/GfxBackend.cpp:141]: Yep, its a small cost to pay though, one texture copy will never be the bottleneck of a renderer in the grand scheme of things.
It allows for WAY cleaner code since otherwise we'd have to have two codepaths, one for the present codepath which waits on the backbuffer acquire and one for the non-compute codepath which submits automatically. From the moment the user uses the backbuffer directly, they will HAVE to border their gpu work with presents (the main issue is that we'd have to be able to know if the user has used the backbuffer in their command list before submitting to be able to apply the correct synchronization, this becomes VERY complicated with bindless...).

Also makes general manipulation of the present textures easier as we're no longer forced into the weird particularities that swapchain textures have (not allowed to be UAVs for example).

And finally, the Vulkan code is WAY easier to implement with the copy so yeah I think its worth it.

File: src/Vex/CommandContext.cpp [hunk 2]
```diff
@@ -300,17 +313,26 @@ void CommandContext::TraceRays(const RayTracingPassDescription& rayTracingPassDe
 
 void CommandContext::Copy(const Texture& source, const Texture& destination)
 {
+    if (source.handle == destination.handle)
+    {
+        VEX_LOG(Fatal, "Cannot copy a texture to itself!");
```
[the remark this answers is not in the document]
Comment (Narvin-Chana) [reply src/Vex/CommandContext.cpp:318]: As mentioned below, I think there is never a valid call to copy that has the same resource as src and dest. Early erroring is  better IMO.

File: src/DX12/RHI/DX12RHI.cpp [hunk 3]
```diff
@@ -228,26 +252,26 @@ void DX12RHI::SubmitAndPresent(std::span<RHICommandList*> commandLists,
...
 
 void DX12RHI::FlushGPU()
 {
+    // Wait for all queues to complete their latest work (the most recently signaled values)
     for (u32 i = 0; i < CommandQueueTypes::Count; ++i)
     {
-        CommandQueueType queueType = static_cast<CommandQueueType>(i);
-        auto& fence = (*fences)[queueType];
-        fence.Flush();
+        auto& fence = (*fences)[static_cast<CommandQueueType>(i)];
```
[the remark this answers is not in the document]
Comment (Narvin-Chana) [reply src/DX12/RHI/DX12RHI.cpp:273]: The issue is that the RHIBase would have to store the command pools, but we've disabled destructor polymorphism on these types.
Maybe a solution would be to have a free function to share the behavior that takes in the RHI version? That would actually work pretty well.

File: src/Vulkan/RHI/VkPipelineState.cpp [hunk 4]
```diff
@@ -255,6 +263,11 @@ void VkRayTracingPipelineState::Compile(const RayTracingShaderCollection& shader
 
 void VkRayTracingPipelineState::Cleanup(ResourceCleanup& resourceCleanup)
 {
+    // Don't cleanup if RTPSO is null.
+    if (!!1)
+    {
+        return;
+    }
```
[the remark this answers is not in the document]
Comment (Narvin-Chana) [reply src/Vulkan/RHI/VkPipelineState.cpp:270]: Just to remind you to add the condition on the RTPSO when you implement the VkRayTracingPSO class.
We had an issue where the first compilation of a PSO would queue in an empty resource for cleanup.

File: src/Vex/GfxBackend.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (alexandreLBarrett) [reply src/Vex/GfxBackend.cpp:141]: Sounds good

File: src/DX12/RHI/DX12RHI.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Narvin-Chana) [reply src/DX12/RHI/DX12RHI.cpp:273]: Done!

File: src/Vulkan/RHI/VkCommandPool.cpp [hunk 5]
```diff
@@ -2,72 +2,115 @@
...
-
-    // Release underlying command buffers now that they are assuredly no longer in flight.
-    commandBuffers.clear();
+    // Save the state of the newly submitted command lists
+    for (u32 i = 0; i < submits.size(); ++i)
+    {
+        auto& cmdList = submits[i];
+        cmdList->SetState(RHICommandListState::Submitted);
+        cmdList->SetSyncTokens(syncTokens);
+    }
```
[the remark this answers is not in the document]
Comment (Narvin-Chana) [reply src/Vulkan/RHI/VkCommandPool.cpp]: Fixed

File: src/DX12/RHI/DX12RHI.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Narvin-Chana) [reply src/DX12/RHI/DX12RHI.cpp:273]: Update we use a std::optional making a foward declaration of this type impossible.
Including the RHIImpl/Fence will create a circular dependency.

I think its ok if we keep it as-is for now. The code duplication isn't too bad.

**Review by alexandreLBarrett — APPROVED** [review]

**Review by alexandreLBarrett — APPROVED** [review]
